tests/tests/test_grouping.py

```diff
@@ -105,48 +101,43 @@ def test_update_device_group(self, standard_setup_two_clients_bootstrapped):
+        alpha = mender_device_group[0]
+        bravo = mender_device_group[1]
 
-        ip_to_device_id = Helpers.ip_to_device_id_map(clients)
-        id_alpha = ip_to_device_id[alpha]
-        id_bravo = ip_to_device_id[bravo]
-        print("ID of alpha host: %s\nID of bravo host: %s" % (id_alpha, id_bravo))
+        ip_to_device_id = Helpers.ip_to_device_id_map(mender_device_group)
+        id_alpha = ip_to_device_id[alpha.host_string]
+        id_bravo = ip_to_device_id[bravo.host_string]
+        logger.info("ID of alpha host: %s" % id_alpha)
+        logger.info("ID of bravo host: %s" % id_bravo)
 
-        ret = execute(Helpers.get_passive_partition, hosts=clients)
-        pass_part_alpha = ret[alpha]
-        pass_part_bravo = ret[bravo]
+        # TODO: paralize these using fabric.group.ThreadingGroup once we upgrade to Python 3
+        pass_part_alpha = Helpers.get_passive_partition(alpha)
+        pass_part_bravo = Helpers.get_passive_partition(bravo)
 
         inv.put_device_in_group(id_alpha, "Update")
 
         reboot = { alpha: None, bravo: None }
         host_ip = standard_setup_two_clients_bootstrapped.get_virtual_network_host_ip()
-        with Helpers.RebootDetector(host_ip, alpha) as reboot[alpha], Helpers.RebootDetector(host_ip, bravo) as reboot[bravo]:
-
-            deployment_id, expected_image_id = common_update_procedure(conftest.get_valid_image(),
-                                                                       devices=[id_alpha])
-
-            @parallel
-            def verify_reboot_performed_for_alpha_only():
-                if env.host_string == alpha:
-                    reboot[alpha].verify_reboot_performed()
-                elif env.host_string == bravo:
-                    # Extra long wait here, because a real update takes quite a lot
-                    # of time.
-                    reboot[bravo].verify_reboot_not_performed(300)
-                else:
-                    raise Exception("verify_reboot_performed_for_alpha_only() called with unknown host")
-
-            execute(verify_reboot_performed_for_alpha_only, hosts=clients)
-
-        ret = execute(Helpers.get_passive_partition, hosts=clients)
-        assert ret[alpha] != pass_part_alpha
-        assert ret[bravo] == pass_part_bravo
-        ret = execute(Helpers.get_active_partition, hosts=clients)
-        assert ret[alpha] == pass_part_alpha
-        assert ret[bravo] != pass_part_bravo
+        with Helpers.RebootDetector(alpha, host_ip) as reboot[
+            alpha
+        ], Helpers.RebootDetector(bravo, host_ip) as reboot[bravo]:
+
+            deployment_id, expected_image_id = common_update_procedure(
+                conftest.get_valid_image(), devices=[id_alpha]
+            )
+
+            reboot[alpha].verify_reboot_performed()
+            # Extra long wait here, because a real update takes quite a lot
+            # of time.
+            reboot[bravo].verify_reboot_not_performed(300)
```

Comment:
Now that these are not parallelized anymore, a small optimization you could try is to call these two in the opposite order. The thing is that `verify_reboot_not_performed` will always wait the entire interval, whereas `verify_reboot_performed` will progress while you're doing other things, so it should return immediately if you call it as number two and `verify_reboot_not_performed` has already waited for a long time.

Reply:
Good idea.